### PR TITLE
show concept path (i.s.o. columnid) in GridView column header tool tip

### PR DESCRIPTION
**Pull request by wimvanderlinden**
@forus This pull request provides a solution to the JIRA issue FT-1270 (https://jira.ctmmtrait.nl/browse/FT-1270, TraIT Foundation Team).

### CODE REVIEW

**Comment by Terry-Weymouth** [comment]
There is another solution, in the code somewhere, I believe associated with the export column headings, that creates the "minimal distinguishing path" as a unique column id. Would that concept also apply here?

**Comment by wimvanderlinden** [comment]
@Terry-Weymouth That might even be a better solution, provided that the "minimal distinguishing path" still is informative to the user and does not contain any of the non-alpha-numeric characters which I had to replace to prevent the grid view from not properly functioning anymore. This character substitution potentially destroys the uniqueness of the path. Can you be more specific about where I can find this piece of code? I certainly would like to make use of that concept if possible.

**Comment by wimvanderlinden** [comment]
@Terry-Weymouth Explicitly passing the text to be shown in the column header tool tip seems like a better and more transparent approach to solve the issue with the non-information in the tool tip for a GridView column header.

File: grails-app/services/I2b2HelperService.groovy [hunk 1]
```diff
@@ -755,15 +755,24 @@ class I2b2HelperService {
...
+         * showing the full concept path in the tooltip is much more informative.
+         * As no tooltip text is passed on to the GridView code, the value of the string columnid is used
+         * and shown as the tooltip text when hoovering over the column header in GridView.
+         * Explicitly passing a tooltip text to the GridView code removes the necessity to use this columnid value.
+         * Removal of some undesired non-alpha-numeric characters from tooltip string
+         * prevents display errors in GridView (drop down menu, columns not showing or cells not being filled).
+         */
         String columnid = concept_key.encodeAsSHA1()
         String columnname = getColumnNameFromKey(concept_key).replace(" ", "_")
+        String columntooltip = keyToPath(concept_key).replace(" ", "_").replace(".","_").replace("[","_")
```
Comment (forus) [inline grails-app/services/I2b2HelperService.groovy]: @wimvanderlinden What about replacing `]` symbol?

Comment (forus) [inline grails-app/services/I2b2HelperService.groovy]: I think it'd be hard to come up with "black list" of symbols. Wouldn't it be easier to replace all non non-word characters? Something like this: `keyToPath(concept_key).replaceAll('[^a-zA-Z_0-9\\\\]+', '_')`. Note we do not replace backslash here `\\\\`.

**Comment by wimvanderlinden** [comment]
@forus 
The ] symbol is not replaced because the presence of this symbol in the path doesn't introduce errors in the Grid View presentation.
Although I tested this with paths containing most of the keyboard symbols (at least the symbols which are accepted during clinical upload), indeed it is difficult to come up and maintain such a black list. Your suggestion to replace all non-preferred character sequences with the _ symbol makes the code more robust. Although potentially, different paths might show up with the same name in the tooltip. But this is already true replacing specific symbols only.
I adjusted the code as you proposed, but also left the hyphen character untouched by the substitution.

**Pull request merged**